### ISSUE TITLE
Add support for XDM-AUTHORIZATION-1

### ISSUE THREAD

**Issue by psychon** [psychon/x11rb#206]
This is split out from #196.

`MIT-MAGIC-COOKIE-1` is easy to implement: just send the secret to the server (unencrypted, so anyone listening now knows the secret).

`XDM-AUTHORIZATION-1` tries to be secure by using some crypto based on the current time and some other things. I haven't really found the actual crypto yet, but given its ages, I doubt that it is good crypto.

Anyway, for now I will only implement the simple case. This issue tracks the complicated case.

**Comment by psychon** [comment]
Some more data points on how to implement this:
- https://www.x.org/releases/X11R7.7/doc/libXdmcp/xdmcp.html#Authentication
  - This uses DES, so needs something like https://crates.io/crates/des as a new (optional?) dependency
- `compute_auth()` in `libxcb/src/xcb_auth.c` computes the input to the algorithm
  - A nonce is used; this is basically an atomic int that is incremented on each use
  - The authentication data is composed / concatenated out of:
    - The first 8 bytes of data from `~/.xauthority`
    - For `AF_INET` sockets: the ip address and port (in network byte order, I guess)
    - For `AF_INET6`: If it is a V4 mapped address: See above; otherwise `0.0.0.0:0` is used (yay! that must be really secure)
    - For `AF_UNIX`: `u32::max_value() - next_nonce()` is interpreted as an IPV4 address and the port is the PID
    - Next, `htonl(time(0))` is appended.
    - All of this is padded with `0` bytes up to length 192 bits / 24 bytes
  - All of this is passed into `XdmcpWrap` as input
  - The `wrapper` argument (could this be the DES key?) of `XdmcpWrap` points to the value from `~/.Xauthority`, but the first 8 bytes are skipped (they were already used above)
  - The output of this is the authentication data (the function is actually called with `input == output`
- No idea what `XdmcpWrap` then actually does. It does not seem to map to the spec in a clear way and... dunno

Basically, this seems to be some complicated magic that I uses crypto that is most likely not secure and somehow I feel all of this is not used these days much. Let's wait for user requests for an implementation (and perhaps some instructions on how to actually test this).

**Comment by psychon** [comment]
xtruss seems to contain some code that can handle XDM auth. I am just not sure what exactly it does, but it seems to be self-contained.
